File: test/mock-llm.js
/**
 * Minimal OpenAI-compatible mock LLM for end-to-end testing without a real
 * API key. Scripted behavior:
 *   - If the last message is not a tool result: call `execute` with an echo.
 *   - If a tool result is present: stream a short final text answer.
 * Run: bun test/mock-llm.js  (listens on :8901)
 */
const PORT = Number(process.env.MOCK_PORT ?? 8901);
// per-word delay when streaming the final answer; raise it (e.g. 800) to
// keep a run in-flight long enough to exercise detach/reattach flows
const DELAY = Number(process.env.MOCK_DELAY_MS ?? 30);

function sse(obj) {
  return `data: ${JSON.stringify(obj)}\n\n`;
}

// one id per completion, shared by all chunks of that completion
function makeChunkFactory() {
  const id = "chatcmpl-" + Math.random().toString(36).slice(2);
  return function chunk(delta, finish = null) {
    return {
      id,
      object: "chat.completion.chunk",
      created: Math.floor(Date.now() / 1000),
      model: "mock-model",
      choices: [{ index: 0, delta, finish_reason: finish }],
    };
  };
}

Bun.serve({
  port: PORT,
  async fetch(req) {
    const url = new URL(req.url);
    if (!url.pathname.endsWith("/chat/completions")) {
      return new Response("not found", { status: 404 });
    }
    const body = await req.json();
    const messages = body.messages ?? [];
    const hasToolResult = messages.some((m) => m.role === "tool");
    const usage = { prompt_tokens: 10, completion_tokens: 10, total_tokens: 20 };

    if (body.stream === false) {
      const message = hasToolResult
        ? { role: "assistant", content: "命令已执行，见工具结果。" }
        : {
            role: "assistant",
            content: "",
            tool_calls: [
              {
                id: "call_mock_1",
                type: "function",
                function: {
                  name: "execute",
                  arguments: JSON.stringify({ command: "echo hello-from-mock-agent" }),
                },
              },
            ],
          };
      return Response.json({
        id: "chatcmpl-" + Math.random().toString(36).slice(2),
        object: "chat.completion",
        created: Math.floor(Date.now() / 1000),
        model: "mock-model",
        choices: [
          {
            index: 0,
            message,
            finish_reason: hasToolResult ? "stop" : "tool_calls",
          },
        ],
        usage,
      });
    }

    const stream = new ReadableStream({
      async start(controller) {
        const chunk = makeChunkFactory();
        const enc = new TextEncoder();
        const push = (s) => controller.enqueue(enc.encode(s));

        if (!hasToolResult) {
          // stream a tool call to `execute`
          push(sse(chunk({ role: "assistant", content: "" })));
          push(
            sse(
              chunk({
                tool_calls: [
                  {
                    index: 0,
                    id: "call_mock_1",
                    type: "function",
                    function: { name: "execute", arguments: "" },
                  },
                ],
              })
            )
          );
          const args = JSON.stringify({ command: "echo hello-from-mock-agent" });
          for (const piece of [args.slice(0, 12), args.slice(12)]) {
            push(
              sse(
                chunk({
                  tool_calls: [{ index: 0, function: { arguments: piece } }],
                })
              )
            );
          }
          push(sse(chunk({}, "tool_calls")));
        } else {
          push(sse(chunk({ role: "assistant", content: "" })));
          for (const word of ["命令", "已执行", "，输出", "见上方", "工具结果。"]) {
            push(sse(chunk({ content: word })));
            await new Promise((r) => setTimeout(r, DELAY));
          }
          push(sse(chunk({}, "stop")));
        }
        // usage chunk (as sent by OpenAI with stream_options.include_usage)
        push(
          sse({
            id: "chatcmpl-" + Math.random().toString(36).slice(2),
            object: "chat.completion.chunk",
            created: Math.floor(Date.now() / 1000),
            model: "mock-model",
            choices: [],
            usage,
          })
        );
        push("data: [DONE]\n\n");
        controller.close();
      },
    });

    return new Response(stream, {
      headers: {
        "Content-Type": "text/event-stream",
        "Cache-Control": "no-cache",
      },
    });
  },
});

console.log(`mock LLM listening on http://127.0.0.1:${PORT}/v1`);

File: test/mock_llm.py
"""Minimal OpenAI-compatible mock LLM for end-to-end testing without a real
API key. Scripted behavior:
  - If the last message is not a tool result: call `execute` with an echo.
  - If a tool result is present: stream a short final text answer.
Run: uv run python test/mock_llm.py  (listens on :8901; see --help)
"""

import argparse
import asyncio
import json
import time
import uuid

import uvicorn
from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse, StreamingResponse

# per-word delay (seconds) when streaming the final answer; raise it (e.g.
# --delay-ms 800) to keep a run in-flight long enough to exercise
# detach/reattach flows
settings = {"delay": 0.03}

app = FastAPI()

USAGE = {"prompt_tokens": 10, "completion_tokens": 10, "total_tokens": 20}


def sse(obj) -> str:
    return f"data: {json.dumps(obj, ensure_ascii=False)}\n\n"


def make_chunk_factory():
    cid = "chatcmpl-" + uuid.uuid4().hex[:12]

    def chunk(delta, finish=None):
        return {
            "id": cid,
            "object": "chat.completion.chunk",
            "created": int(time.time()),
            "model": "mock-model",
            "choices": [{"index": 0, "delta": delta, "finish_reason": finish}],
        }

    return chunk


@app.post("/{path:path}")
async def completions(path: str, request: Request):
    if not path.endswith("chat/completions"):
        return JSONResponse({"error": "not found"}, status_code=404)
    body = await request.json()
    messages = body.get("messages") or []
    has_tool_result = any(m.get("role") == "tool" for m in messages)

    if not body.get("stream"):  # 与真实 API 一致：未显式要求流式则返回 JSON
        if has_tool_result:
            message = {"role": "assistant", "content": "命令已执行，见工具结果。"}
        else:
            message = {
                "role": "assistant",
                "content": "",
                "tool_calls": [
                    {
                        "id": "call_mock_1",
                        "type": "function",
                        "function": {
                            "name": "execute",
                            "arguments": json.dumps({"command": "echo hello-from-mock-agent"}),
                        },
                    }
                ],
            }
        return {
            "id": "chatcmpl-" + uuid.uuid4().hex[:12],
            "object": "chat.completion",
            "created": int(time.time()),
            "model": "mock-model",
            "choices": [
                {
                    "index": 0,
                    "message": message,
                    "finish_reason": "stop" if has_tool_result else "tool_calls",
                }
            ],
            "usage": USAGE,
        }

    async def gen():
        chunk = make_chunk_factory()
        if not has_tool_result:
            # stream a tool call to `execute`
            yield sse(chunk({"role": "assistant", "content": ""}))
            yield sse(
                chunk(
                    {
                        "tool_calls": [
                            {
                                "index": 0,
                                "id": "call_mock_1",
                                "type": "function",
                                "function": {"name": "execute", "arguments": ""},
                            }
                        ]
                    }
                )
            )
            args = json.dumps({"command": "echo hello-from-mock-agent"})
            for piece in (args[:12], args[12:]):
                yield sse(chunk({"tool_calls": [{"index": 0, "function": {"arguments": piece}}]}))
            yield sse(chunk({}, "tool_calls"))
        else:
            yield sse(chunk({"role": "assistant", "content": ""}))
            for word in ["命令", "已执行", "，输出", "见上方", "工具结果。"]:
                yield sse(chunk({"content": word}))
                await asyncio.sleep(settings["delay"])
            yield sse(chunk({}, "stop"))
        # usage chunk (as sent by OpenAI with stream_options.include_usage)
        yield sse(
            {
                "id": "chatcmpl-" + uuid.uuid4().hex[:12],
                "object": "chat.completion.chunk",
                "created": int(time.time()),
                "model": "mock-model",
                "choices": [],
                "usage": USAGE,
            }
        )
        yield "data: [DONE]\n\n"

    return StreamingResponse(gen(), media_type="text/event-stream")


def main() -> None:
    parser = argparse.ArgumentParser(description="OpenAI-compatible mock LLM")
    parser.add_argument("--port", type=int, default=8901)
    parser.add_argument("--delay-ms", type=float, default=30)
    args = parser.parse_args()
    settings["delay"] = args.delay_ms / 1000
    uvicorn.run(app, host="127.0.0.1", port=args.port, log_level="warning")


if __name__ == "__main__":
    main()
